File: active_subspaces/Matlab_Version/Python_Testing/Analytic_Function/test_script_subspaces.py
import utils_analytic
import numpy as np
import active_subspaces
import matplotlib.pyplot as plt

# Set quantity of interest and input labels
QofI = 'U_avg'
in_labels = ['mu', 'rho', 'dpdx', 'eta', 'B0']

# Set the number of parameter (m) and active subspace dimension (n)
m, n = 5, 4

# Obtain M input samples drawn uniformly from the input space
M = 100
X_norm = 2*np.random.rand(M, m) - 1
X_phys = utils_analytic.normalized_to_physical(X=X_norm)

# Run simulation on sample inputs
f = utils_analytic.func(X_phys, QofI=QofI)
df_phys = utils_analytic.grad(X_phys, QofI=QofI)

# Scale data so parameters are in [-1,1]
df_norm = utils_analytic.physical_to_normalized(df=df_phys)

# Compute the active/inactive subspaces
# NOTE: The gradient is normalized for this computation
sub = active_subspaces.subspaces.Subspaces()
sub.compute(df_norm/np.linalg.norm(df_norm, axis=1).reshape((df_norm.shape[0], 1)))

# Rewrite the active/inactive subspace variables to be n-dimensional
sub.W1, sub.W2 = sub.eigenvectors[:,:n], sub.eigenvectors[:,n:]
sub.W1 = sub.W1.reshape(m, n)
sub.W2 = sub.W2.reshape(m, m-n)

# Define the active/inactive variables
Y, Z = np.dot(X_norm, sub.W1), np.dot(X_norm, sub.W2)

# Plot the active subspace info
in_labels = ['mu', 'rho', 'dpdx', 'eta', 'B0']
active_subspaces.utils.plotters.eigenvalues(sub.eigenvalues, e_br=sub.e_br, out_label=QofI)
active_subspaces.utils.plotters.eigenvectors(sub.W1, in_labels=in_labels, out_label=QofI)
active_subspaces.utils.plotters.subspace_errors(sub.sub_br, out_label=QofI)
if (n <= 2):
    active_subspaces.utils.plotters.sufficient_summary(Y, f, out_label=QofI)

File: active_subspaces/Matlab_Port/test_script_Subspaces.m
clear variables
close all
clc

addpath 'Subspaces' 'Plotters'

%% Analytic Function Testing - EITHER THIS OR 'Data Set Testing' MUST BE COMMENTED OUT
% 
% m = 5;
% M = 100;
% X_norm = 2*rand(M, m) - 1;
% 
% input_ranges = [0.5*ones(1, m); 3*ones(1, m)];
% 
% X_phys = zeros(size(X_norm));
% for i = 1:M
%     % Scale inputs so that they are on the interval [0.5, 3]
%     X_phys(i, :) = (X_norm(i, :) + 1).*(input_ranges(2, :) - input_ranges(1, :))/2 + input_ranges(1, :);
% end
% 
% [f, df] = analytic_function(X_phys);
% 
% for i = 1:M
%     % Scale gradient to reflect inputs drawn from [-1, 1]^m domain.
%     df(i, :) = df(i, :).*(input_ranges(2, :) - input_ranges(1, :))/2;
%     
%     % Normalize gradient.  Not sure if this is really important
%     df(i, :) = df(i, :)/norm(df(i, :));
% end

%% Data Set Testing

data = importdata('test_data.dat');

% Set m=80 for full data set.
m = 80; n = 2;
X_phys = data(1:m, 1:6);
f = data(1:m, 7);
df = data(1:m, 8:13);

input_ranges = [0.5, 5e-7, 2.5, 0.1, 5e-7, 0.1; ...
                  2, 5e-6, 7.5,  10, 5e-6,  10];

[M, m] = size(df);
for i = 1:M
    % Scale inputs so that they are on the interval [-1, 1]
    X_norm(i, :) = 2*(X_phys(i, :) - input_ranges(1, :))./(input_ranges(2, :) - input_ranges(1, :)) - 1;
    
    % Apply scaling to gradient.
    df(i, :) = df(i, :).*(input_ranges(2, :) - input_ranges(1, :))/2;
    
    % Normalize gradient.  Not sure if this is really important
    df(i, :) = df(i, :)/norm(df(i, :));
end

%% Compute active subspace

% Compute and plot active subspace using gradients approximated by
% local_linear_gradients functions.
sub = compute(df);

% Plot results
eigenvalues(sub.eigenvalues, sub.e_br)
eigenvectors(sub.eigenvectors(:, 1:4))
subspace_errors(sub.sub_br)
sufficient_summary(X_norm*sub.eigenvectors(:, 1:2), f)